cEP-0025.md

```diff
@@ -0,0 +1,306 @@
+                        diff.modify_line(lineno, replacement)
+                        break
+
+        return diff
+
+    def run(self, filename, file,
+            dependency_results=dict()
+            ):
+        """
+        Uses PyFlakesASTBear to get only FutureImportation
+        """
+        # Preprocess all the lines containing __future__ imports
+        affected_lines = set()
+        # Get the maximum column offset of FutureImportation for a line
+        max_column_for_a_line = dict()
+        for result in dependency_results.get(PyFlakesASTBear.name, []):
+            for node in result.get_nodes(result.module_scope[0],
+                                         FutureImportation):
+                lineno = result.get_node_location(node)[0]
+                col_offset = result.get_node_location(node)[2]
+                if lineno in max_column_for_a_line:
+                    max_column_for_a_line[lineno] = max(
+                                max_column_for_a_line[lineno],
+                                col_offset)
+                else:
+                    max_column_for_a_line[lineno] = col_offset
+                affected_lines.add(lineno)
+
+        for line in affected_lines:
+            corrected = self.remove_future_imports(
+                        file, line, max_column_for_a_line[lineno],
+                        affected_lines
+                        )
+            yield Result.from_values(
+                origin=self,
+                message=('Future import/imports found'),
+                file=filename,
+                diffs={filename: corrected},
+                line=line)
+```
+
+## A generic implementation for detecting `__future__` imports
+
+This implementation doesn't requires coala to be installed and is
+designed so that it can be used in other tools like flake8. This
+implementation will be housed in `pyflakes-generic-plugins`.
+
+Here is a prototype for the NoFutureImport that uses pyflakes-AST:
+
+```python
+# Imports not mentioned to maintain brevity
+__version__ = '0.1'
+
+CODE = 'F482'
+
+
+class NoFutureImport(object):
+    name = 'no_future'
+    version = __version__
+
+    def __init__(self, tree, filename):
+        self.tree = tree
+        self.filename = filename
+        self._checker = None
+        self._module_scope = None
+
+    @property
+    def checker(self):
+        if self._checker is None:
+            self._checker = Checker(self.tree, self.filename)
+        return self._checker
+
+    @checker.setter
+    def checker(self, checker):
+        self.checker = checker
+
+    @property
+    def module_scope(self):
+        if self._module_scope is None:
+            self._module_scope = list(filter(lambda scope: isinstance(scope,
+                                                            ModuleScope),
+                                             self.checker.deadScopes))[0]
+        return self._module_scope
+
+    def set_pyflakes_data(self, checker):
```

Comment:
This is unnecessary now ?

Reply:
Will flake8 use `setattr(foo, 'checker', instance)` to initialize the plugin with checker class? Or should we expose this function so that flake8 can use `set_pyflakes_data(instance)` to provide pyflakes AST?

Reply:
flake8 could use `foo.checker` , because it is a property